codec-dns/src/main/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoder.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.CorruptedFrameException;
+import io.netty.util.CharsetUtil;
+
+public class DefaultDnsRecordDecoder implements DnsRecordDecoder {
+
+    @Override
+    public DnsQuestion decodeQuestion(ByteBuf in) throws Exception {
+        String name = decodeName(in);
+        DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        DnsRecordClass qClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        return new DefaultDnsQuestion(name, type, qClass);
+    }
+
+    @Override
+    public <T extends DnsRecord> T decodeRecord(ByteBuf in) throws Exception {
+        final String name = decodeName(in);
+        final DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        final DnsRecordClass aClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        final long ttl = in.readUnsignedInt();
+        final int length = in.readUnsignedShort();
+        final int offset = in.readerIndex();
+
+        @SuppressWarnings("unchecked")
+        T record = (T) decodeRecord(name, type, aClass, ttl, in, offset, length);
```

Comment:
Yeah, it's sufficient to just return `DnsRecord`, and I think it doesn't really matter because it's used only by the DNS decoder.  However, it could potentially be used by some user code which uses this class to decode a certain DNS record from a `ByteBuf`.  The caller might know the exact type of the decoded message, and then this return type signature might be useful, just like the `EmbeddedChannel.readInbound/Outbound()`.